proxy_agent/src/proxy/proxy_server.rs

```diff
@@ -0,0 +1,694 @@
+                    format!("Failed to get frame data: {:?}", e),
+                );
+                Bytes::new()
+            }
+        };
+
+        Frame::data(frame)
+    });
+    let mut response = Response::from_parts(head, frame_stream.boxed());
+
+    // insert default x-ms-azure-host-authorization header to let the client know it is through proxy agent
+    response.headers_mut().insert(
+        HeaderName::from_static(constants::AUTHORIZATION_HEADER),
+        HeaderValue::from_static("value"),
+    );
+
+    // print the response headers
+    Connection::write(
+        connection.id,
+        format!(
+            "Response headers: {:?}",
+            response.headers().iter().collect::<Vec<_>>()
+        ),
+    );
+
+    log_connection_summary(&connection, response.status(), shared_state.clone());
+    Ok(response)
+}
+
+fn log_connection_summary(
+    connection: &ConnectionContext,
+    response_status: StatusCode,
+    shared_state: Arc<Mutex<SharedState>>,
+) {
+    let elapsed_time = connection.now.elapsed();
+    let claims = match &connection.claims {
+        Some(c) => c.clone(),
+        None => Claims::empty(),
+    };
+
+    let summary = ProxySummary {
+        id: connection.id,
+        userId: claims.userId,
+        userName: claims.userName.to_string(),
+        userGroups: claims.userGroups.clone(),
+        clientIp: claims.clientIp.to_string(),
+        processFullPath: claims.processFullPath.to_string(),
+        processCmdLine: claims.processCmdLine.to_string(),
+        runAsElevated: claims.runAsElevated,
+        method: connection.method.to_string(),
+        url: connection.url.to_string(),
+        ip: connection.ip.to_string(),
+        port: connection.port,
+        responseStatus: response_status.to_string(),
+        elapsedTime: elapsed_time.as_millis(),
+    };
+    if let Ok(json) = serde_json::to_string(&summary) {
+        event_logger::write_event(
+            event_logger::INFO_LEVEL,
+            json,
+            "log_connection_summary",
+            "proxy_listener",
+            Connection::CONNECTION_LOGGER_KEY,
+        );
+    };
+    agent_status_wrapper::add_one_connection_summary(shared_state, summary, false);
+}
+
+// We create some utility functions to make Empty and Full bodies
+// fit our broadened Response body type.
+fn empty_response(status_code: StatusCode) -> Response<BoxBody<Bytes, hyper::Error>> {
+    let empty = Empty::<Bytes>::new()
+        .map_err(|never| match never {})
+        .boxed();
+
+    let mut response = Response::new(empty);
+    *response.status_mut() = status_code;
+
+    response
+}
+
+async fn handle_request_with_signature(
+    connection: ConnectionContext,
+    request: Request<hyper::body::Incoming>,
+    shared_state: Arc<Mutex<SharedState>>,
+) -> Result<Response<BoxBody<Bytes, hyper::Error>>, hyper::Error> {
+    let (head, body) = request.into_parts();
+    let whole_body = match body.collect().await {
```

Comment:
There doesn't appear to be any limit on the request size. Even though it looks like this happens after the authentication check, accepting arbitrarily large requests is probably not a good idea. There's a bunch of pre-built middleware for this kind of thing in [tower](https://tower-rs.github.io/tower/tower/) which should integrate nicely with Hyper (you can probably also implement authentication as a Tower service, as well).

Reply:
Good idea to keep GPA service not OOM. Let me discuss with the host team to decide the size limitation.